### ISSUE TITLE
[Binary Releases] Master issue to deliver the infrastructure for auto-generating RPM and DEB packages for our users

### ISSUE THREAD

**Issue by paulkatsoulakis** [netdata/netdata#5963]
###### Introduction
  Having seen the need for our users to be able to install netdata hustle-free, without having to worry about scripts and sources, we decided to proceed with this project to help our users get the latest netdata on their systems with zero effort. We are building our first infrastructure that will provide netdata with the auto-generation of RPM and DEB files to cover most distributions netdata is able to run at the moment.
  This is just an initial project that will provide us with a baseline implementation.
Once this is complete a constant re-iteration of improvements and adjustments will follow, so that we can guarantee our infrastructure is up to standards with the market trends around package management.

Note: There was #5775 that took place prior to deciding the plan for this story, were we investigate all 

###### Tasks
Here's the activities that need to take place in order for us to get our milestone complete.
We are targeting to start with an x86-64 version 

- [ ] Document roadmap and status for packaging and distribution support #6036
- [x] Establish Debian/Ubuntu repository servers and software #5965 
- [x] Establish RPM repository servers and software #5966
- [x] Create a script to containerise the RPM build process #5967
- [x] Create a script that will containerise the DEB build process #5968
- [x] Optimise netdata.spec file, sync parts of the `netdata-installer.sh` steps to the spec file #5969
- [ ] [Placeholder] Optimize DEB build process (post-creation step to further optimise initial artifacts)
- [ ] Establish gentoo portage repository infrastructure #5979
- [ ] Make go.d plugin an independent package #6367 6367

Low priority, for later:
- [ ] Remove `v` from version #5983

##### Expected Behaviour
At the end of this story, netdata will provide its own repositories for DEP and RPM packages.
Our users will be able to setup their whatever package managers with our own repository servers to retrieve latest netdata binaries directly from us.

**Comment by woosley** [comment]
since there is a spec file, is there already RPM package availabe somewhere?

**Comment by paulkatsoulakis** [comment]
> since there is a spec file, is there already RPM package availabe somewhere?

Hi @woosley, thanks for reaching out!

We already have several users from our community that has published and maintain an RPM for various distributions. If you check your package management utility you might find a version of netdata already available for install, but is likely not to be the latest one.

For example on a centOS 7 if you install the epel-release repo:
```
[root@XXXXXXXX netdata]# yum list netdata
Loaded plugins: fastestmirror, ovl
Determining fastest mirrors
epel/x86_64/metalink                                                                                                                                                                |  23 kB  00:00:00     
 * base: mirror.strencom.net
 * epel: anorien.csc.warwick.ac.uk
 * extras: mirrors.vooservers.com
 * updates: mirror.econdc.com
base                                                                                                                                                                                | 3.6 kB  00:00:00     
epel                                                                                                                                                                                | 4.7 kB  00:00:00     
extras                                                                                                                                                                              | 3.4 kB  00:00:00     
updates                                                                                                                                                                             | 3.4 kB  00:00:00     
(1/7): base/7/x86_64/group_gz                                                                                                                                                       | 166 kB  00:00:00     
(2/7): extras/7/x86_64/primary_db                                                                                                                                                   | 201 kB  00:00:00     
(3/7): epel/x86_64/group_gz                                                                                                                                                         |  88 kB  00:00:01     
(4/7): epel/x86_64/updateinfo                                                                                                                                                       | 992 kB  00:00:02     
(5/7): updates/7/x86_64/primary_db                                                                                                                                                  | 4.2 MB  00:00:05     
(6/7): epel/x86_64/primary_db                                                                                                                                                       | 6.7 MB  00:00:14     
(7/7): base/7/x86_64/primary_db                                                                                                                                                     | 6.0 MB  00:00:16     
Available Packages
netdata.x86_64                                                                                      1.14.0-1.el7                                                                                       epel
```

**Comment by paulkatsoulakis** [comment]
Regarding the distributions we choose to start our support with, here's a report demonstrating the top versions our community uses:

```
alpine  | 57165
amzn   | 1688
arch     | 993
centos | 12226
coreos | 1908
debian | 97567
fedora  | 169
gentoo | 188
nethserver | 185
openwrt_koolshare_mod | 10173
raspbian      | 923
rhel              | 618
ubuntu         | 35180
unknown      | 3014

Grand Total | 221997
```

Moving forward for this first iteration for the binary releases, we plan to support DEB and RPM packages to cover the following distributions:

Debian
Ubuntu
Alpine
Centos
Fedora
Redhat

And any other RPM or DEB compatible distribution going foward.
This will not be the end though, we plan to support the majority of the distributions, this is just to touch base and cover the larger groups in the first round

**Comment by Ferroin** [comment]
Nitpick: Alpine is not a DEB or RPM based platform, they use a custom package format (though they shouldn't be hard to deal with).

Also, on the note of Gentoo (and by extension it's derivatives), I can help with handling of that if desired.  I've been toying with just taking the time to set up a Netdata Portage overlay anyway as the official ebuild that's already in Portage _still_ doesn't support the Go plugin yet.  On that side of things, there should be next to no changes required in the Netdata build system or code itself (that's part of the beauty of Gentoo, it just leverages what's already there for building the software).

**Comment by paulkatsoulakis** [comment]
> Nitpick: Alpine is not a DEB or RPM based platform, they use a custom package format (though they shouldn't be hard to deal with).
> 
> Also, on the note of Gentoo (and by extension it's derivatives), I can help with handling of that if desired. I've been toying with just taking the time to set up a Netdata Portage overlay anyway as the official ebuild that's already in Portage _still_ doesn't support the Go plugin yet. On that side of things, there should be next to no changes required in the Netdata build system or code itself (that's part of the beauty of Gentoo, it just leverages what's already there for building the software).

Yep, correct! I skipped to mention that, i investigated the amount of Alpine being docker because if it wasn't containers, i would have to consider separate infrastructure for APK building and publishing!

Regarding gentoo, given the low usage is not going to be prioritised at this stage.
By all means your contribution though would be highly appreciated.

You can post any instructions on the master issue and i 'll get another subtask out to implement whatever is needed.

**Comment by Ferroin** [comment]
> Regarding gentoo, given the low usage is not going to be prioritised at this stage.
> By all means your contribution though would be highly appreciated.
>
> You can post any instructions on the master issue here and i 'll get another subtask out to implement whatever is needed.

For the Gentoo side of things it comes down to:

* Ensuring we have a way to download source tarballs for each release (I'm pretty sure we already have this covered, I'm just not sure if we're providing it through anything other than GitHub).
* Providing a Portage overlay for the Netdata ebuilds (packages in non-Gentoo terms).  This can be trivially hosted on Github, and is largely just a collection of special scripts that get parsed by emerge to figure out how to build the software and what files to install where (in essence, spec files, just a different format).  Check https://github.com/Ferroin/ahferroin7-overlay for a trivial example of what this might look like.
* Get the overlay on the central list of Portage overlays provided by Gentoo (this shouldn't be hard, should only need to be done once, and mostly just makes life easier for people who want to use it).
* Keeping the ebuilds sufficiently up-to-date to allow use of new features.
* Creating a new ebuild for each version of Netdata, the Go plugin, and any other extra packages, and cleaning up ebuilds for old versions.
* Verifying that the ebuilds work (this is trivial to automate, you just spin up a Gentoo container, and step through installing each of the packages).

I can provide all the initial work for the overlay itself (in fact, I can probably have that ready this week if I don't get swamped with stuff at work), but I think it would be best to have it hosted under the Netdata organization instead of some specific user.  I can also help with the third part (other than new optional core data collectors, Netdata itself is pretty trivial for that, it will be the Go plugin that gets tricky because every collector is theoretically optional for that).  If we provide 'live' ebuilds (that is, ones which build against the current tip of the upstream master VCS branch), we can easily automate the fourth part too (just copy the live ebuild for the new version, and keep X old versions).

**Comment by Ferroin** [comment]
@paulkatsoulakis I've got the beginnings of a Portage overlay for Netdata mostly ready to go, but I'm going to need feedback on a couple of things before I can really say it's 'ready' to put up anywhere.  Do you want me to open a separate issue for Gentoo support to discuss that so we don't swamp things here with stuff that's specific to Gentoo support?

**Comment by paulkatsoulakis** [comment]
> @paulkatsoulakis I've got the beginnings of a Portage overlay for Netdata mostly ready to go, but I'm going to need feedback on a couple of things before I can really say it's 'ready' to put up anywhere. Do you want me to open a separate issue for Gentoo support to discuss that so we don't swamp things here with stuff that's specific to Gentoo support?

You rock man! Yes, please do open an issue and i will add it on my master list above, i created the entry and waiting for the issue id to add it for reference.

**Comment by Ferroin** [comment]
@paulkatsoulakis Opened as #5979.

**Comment by markusr** [comment]
@paulkatsoulakis 
Do you also have statistics on the CPU architectures?

I think main architectures are: ```amd64```, ```i386``` and ```armhf```.

**Comment by paulkatsoulakis** [comment]
> @paulkatsoulakis
> Do you also have statistics on the CPU architectures?
> 
> I think main architectures are: `amd64`, `i386` and `armhf`.
Hi @markusr, thanks for your input!
Actually from the report i have at hand, the majority of the installations are amd64/x86_64 but we do plan to support multiple architectures nonetheless

**Comment by paulkatsoulakis** [comment]
Overall status update:

* We have a good progress on the project, infrastructure matters are quickly out of the way.
* We have finished the design and implementation on the CI pipeline that will take care the deployment of the produced binaries to the repository infrastructure.

There are still some open issues that will be tracked under #5967 and #5968 mostly related to travis and some to lack of experience on Debian packaging. I haven't updated the issues yet, will update this morning with all the progress that has been made the previous days.

Regarding the release of this master issue, I was targeting for 1.15 to get something ready, but given the issues i 've seen with Travis, i won't risk merging anything yet. I will probably move all items on the next milestone

**Comment by lskillen** [comment]
Hey folks - Saw that one of your users setup a test repository for this on @cloudsmith-io. We'd (I'm a Cloudsmither) be delighted to help support you if you'd like to let us handle all of the infrastructure for distribution. As per the list above, we support the majority of the formats and distributions (e.g. Debian, Ubuntu, CentOs, RedHat, Alpine, etc.) that you require, and we're receptive to adding more as required. If you'd really rather host everything yourself, happy to help with questions instead. :-) Good luck!

**Comment by paulkatsoulakis** [comment]
> Hey folks - Saw that one of your users setup a test repository for this on @cloudsmith-io. We'd (I'm a Cloudsmither) be delighted to help support you if you'd like to let us handle all of the infrastructure for distribution. As per the list above, we support the majority of the formats and distributions (e.g. Debian, Ubuntu, CentOs, RedHat, Alpine, etc.) that you require, and we're receptive to adding more as required. If you'd really rather host everything yourself, happy to help with questions on packaging itself instead. :-) Good luck!
> 
> P.s. Netdata is an awesome piece of kit. As a fervent believer in the Church of Graphs (esp. pretty ones). 😍

Hello @lskillen, glad to see people recognising our effort, it's awesome!
We will definitely be in touch, at the very worst case to collaborate on packaging issues.
cloudsmith.io offering looks quite interesting, we will investigate the available features and decide on an approach to this, probably on one of the upcoming sprints.

Thanks again, we 'll be in touch probably through this issue again!

**Comment by lskillen** [comment]
@paulkatsoulakis Haven't heard from you, so assuming you went on with pkgcloud (no hard feelings!) We'd happily sit alongside another distributor if you need us. There's also native support for Alpine packages and other formats if those help too; not to mention org support. No matter what though, as a user, great to see an automated channel for easier installation. Keep up the great work! :heart_eyes: :tada:

**Comment by prologic** [comment]
As we already have active support for a limited set of RPM and DEB packages _and_ #7826 is actively being; worked which expands our support for binary packages and close to being delivered, I am closing this as "done" as the infra as per the title is mostly there anyway.
